editor/plugins/visual_shader_editor_plugin.cpp

```diff
@@ -440,6 +440,10 @@ void VisualShaderGraphPlugin::add_node(VisualShader::Type p_type, int p_id, bool
 		custom_node->_set_initialized(true);
 	}
 
+	if (!p_just_update) {
+		vsnode->connect("node_ports_removed", callable_mp(editor, &VisualShaderEditor::_node_ports_removed).bind(p_id));
```

Comment:
> But the editor crashes as a result...

```suggestion
		vsnode->connect("node_ports_removed", callable_mp(editor, &VisualShaderEditor::_node_ports_removed).bind(p_id), CONNECT_DEFERRED);
```

I think you should try to use CONNECT_DEFERRED flag to prevent that crash.

Reply:
I'll give it a shot thank you

Reply:
It did fix the crash! I have another issue though: the op_type change and the connections removal are not in the same action, which means that hitting ctrl+z only recreates the connections without reverting to the previous op_type. As far as I can understand, handling removed ports should be done in `_property_changed` function?

Reply:
Then unify it to a single action, (`CONNECT_DEFERRED` will not be needed in that case), remove create_action/commit_action from the `_node_ports_removed` and uses the same action as usual just call it before the `commit_action` of the op_type change function has been called.

Reply:
Yes that's what I was proposing: the `commit_action` gets called in `VisualShaderNodePluginDefaultEditor::_property_changed` so all I need to do is check for invalid connections there.
It means that invalid connections will be checked everytime a property gets called whether this property has an impact on the number of ports or not. On the other hand, it also means that no matter what happens in the future invalid connections will always be removed as they should be